ably/realtime_conn.go

```diff
@@ -593,11 +597,16 @@ func (c *Connection) advanceSerial() {
 func (c *Connection) send(msg *protocolMessage, onAck func(err error)) {
 	hasMsgSerial := msg.Action == actionMessage || msg.Action == actionPresence
 	c.mtx.Lock()
+	// RTP16a - in case of presence msg send, check for connection status and send accordingly
 	switch state := c.state; state {
 	default:
 		c.mtx.Unlock()
 		if onAck != nil {
-			onAck(connStateError(state, nil))
+			if c.state == ConnectionStateClosed {
+				onAck(errClosed)
```

Comment:
why does `closed` need to be special-cased? can't it be handled by `connStateError(state)` like all the others?

Reply:
Actually 
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/state.go#L55-L60 map doesn't contain `errClosed`. 
At the same time, we can't update the map to include `errClosed` since it can be a graceful close from the client side. 
If we update the map with `errClosed`, other tests start failing where we don't expect error when client goes into closed state, like when we call `client.close()` and `conn` state changes to `closed` with error.
This error is specific to our usecase, so I didn't include it into the map

Reply:
> At the same time, we can't update the map to include errClosed since it can be a graceful close from the client side.

I don't follow. If the connection is closed, then most operations you try to do to it (other than reconnect) should fail with a 'connection closed' error, similarly to if it's failed. if connclosed is missing from that then either every place you call is going to need to check if the conn is closed and handle that explicitly, or you have a bug, no?

> E.g. call to client.close(), and conn state changes to closed with error.

No it doesn't because connection state changes don't have errors. They have `reason`s. The reason might be 'the connection was closed', that's not an error, it's telling you the connection is in that state

Reply:
I believe it's called `errorReason` =>
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/realtime_conn.go#L48-L51
and same is copied into reason
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/realtime_conn.go#L1001-L1006

There are tests which doesn't treat connection closed as connection failure as such. I mean ideally they shouldn't.
Because, `client.close()` initiates it and gracefully closes the connection accordingly.

Reply:
Do we have a spec item that specifies to set a reason when client explicitly sends a `close` message and  connection is closed by server